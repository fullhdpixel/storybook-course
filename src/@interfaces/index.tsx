export * from "./Option.d";
export * from "./Rental.d";
