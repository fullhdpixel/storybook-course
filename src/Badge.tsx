import React, { FC } from "react";
import classNames from "classnames";

type BadgeVariant = "gray" | "primary" | "error" | "warning" | "success";
type BadgeSize = "sm" | "md" | "lg";

export interface BadgeProps {
  children: string | React.ReactElement;
  variant: BadgeVariant;
  size?: BadgeSize;
  LeadingIcon?: React.ReactElement;
  TrailingIcon?: React.ReactElement;
}

const BadgeBaseClasses = "inline-flex items-center rounded-2xl font-medium";

const BadgeVariantClasses: Record<BadgeVariant, string> = {
  gray: "bg-gray-100 text-gray-700",
  primary: "bg-primary-50 text-primary-700",
  error: "bg-error-50 text-error-700",
  warning: "bg-warning-50 text-warning-700",
  success: "bg-success-50 text-success-700",
};

const BadgeSizeClasses: Record<BadgeSize, string> = {
  sm: "h-5.5 px-2 text-xs",
  md: "h-6 px-2.5 text-xs",
  lg: "h-7 px-3 text-sm",
};

export const Badge: FC<BadgeProps> = ({
  variant,
  size = "md",
  LeadingIcon,
  TrailingIcon,
  children,
}) => {
  const BadgeVariantClassName = BadgeVariantClasses[variant];
  const BadgeSizeClassName = BadgeSizeClasses[size];

  return (
    <div
      className={classNames(
        BadgeBaseClasses,
        BadgeVariantClassName,
        BadgeSizeClassName,
      )}
    >
      {LeadingIcon ? (
        <LeadingIcon.type {...LeadingIcon.props} className="mr-1.5" />
      ) : null}
      {children}
      {TrailingIcon ? (
        <TrailingIcon.type {...TrailingIcon.props} className="ml-1.5" />
      ) : null}
    </div>
  );
};
