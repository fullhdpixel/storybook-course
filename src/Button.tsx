import React, { FC, ButtonHTMLAttributes } from "react";
import classNames from "classnames";

type ButtonVariant =
  | "primary"
  | "secondary"
  | "secondaryGray"
  | "tertiary"
  | "tertiaryGray";

type ButtonSize = "sm" | "md" | "lg" | "xl" | "2xl";

const ButtonBaseClasses =
  "flex items-center rounded-lg font-medium focus:outline-none";

type ButtonState = "default" | "hover" | "focus" | "disabled";

const ButtonVariantClasses: Record<
  ButtonVariant,
  Record<ButtonState, string>
> = {
  primary: {
    default: "bg-primary-600 dark:bg-gray-800 text-white dark:text-white",
    hover: "hover:bg-primary-700 dark:hover:bg-gray-700",
    focus:
      "focus:ring focus:border-primary-100 dark:focus:border-gray-300 dark:focus:border-opacity-10",
    disabled: "bg-primary-200 text-white",
  },
  secondary: {
    default: "bg-primary-50 dark:bg-gray-800 text-primary-700",
    hover: "hover:bg-primary-100",
    focus:
      "focus:ring focus:border-primary-100 dark:focus:border-gray-300 dark:focus:border-opacity-10",
    disabled: "bg-primary-25 text-primary-300",
  },
  secondaryGray: {
    default: "bg-white text-gray-700 border border-gray-300",
    hover: "hover:bg-gray-50 hover:text-gray-800",
    focus:
      "focus:ring focus:border-primary-100 dark:focus:border-gray-300 dark:focus:border-opacity-10",
    disabled: "border border-gray-200 text-gray-300",
  },
  tertiary: {
    default: "bg-white text-primary-700",
    hover: "hover:bg-primary-50",
    focus:
      "focus:ring focus:border-primary-100 dark:focus:border-gray-300 dark:focus:border-opacity-10",
    disabled: "bg-white text-gray-300",
  },
  tertiaryGray: {
    default: "bg-white text-gray-500",
    hover: "hover:bg-gray-50 hover:text-gray-600",
    focus:
      "focus:ring focus:border-primary-100 dark:focus:border-gray-300 dark:focus:border-opacity-10",
    disabled: "bg-white text-gray-300",
  },
};

const ButtonSizeClasses: Record<ButtonSize, string> = {
  sm: "h-9 px-3.5 text-sm",
  md: "h-10 px-4 text-sm",
  lg: "h-11 px-4.5 text-md",
  xl: "h-12 px-5 text-md",
  "2xl": "h-15 px-7 text-lg",
};

const ButtonIconSizeClasses: Record<ButtonSize, string> = {
  sm: "h-9 w-9",
  md: "h-10 w-10",
  lg: "h-11 w-11",
  xl: "h-12 w-12",
  "2xl": "h-15 w-15",
};

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  children?: string | React.ReactElement;
  className?: string;
  variant: ButtonVariant;
  size?: ButtonSize;
  LeadingIcon?: React.ReactElement;
  TrailingIcon?: React.ReactElement;
  IconOnly?: React.ReactElement;
  disabled?: boolean;
}

// Dark Mode
export const Button: FC<ButtonProps> = ({
  children,
  className,
  variant = "primary",
  size = "md",
  LeadingIcon,
  TrailingIcon,
  IconOnly,
  disabled,
  ...buttonProps
}) => {
  const ButtonVariantClassName = ButtonVariantClasses[variant];
  const ButtonIconSizeClassName = ButtonIconSizeClasses[size];

  return (
    <button
      {...buttonProps}
      className={classNames(ButtonBaseClasses, className, {
        [ButtonSizeClasses[size]]: !IconOnly,
        [classNames(ButtonIconSizeClassName, "justify-center")]: IconOnly,
        [classNames(
          ButtonVariantClassName.default,
          ButtonVariantClassName.hover,
          ButtonVariantClassName.focus,
        )]: !disabled,
        [classNames(ButtonVariantClassName.disabled, "cursor-not-allowed")]:
          disabled,
      })}
    >
      {LeadingIcon ? (
        <LeadingIcon.type
          {...LeadingIcon.props}
          className={classNames({
            "mr-2": size !== "2xl",
            "mr-3": size === "2xl",
          })}
        />
      ) : null}
      {children}
      {IconOnly ? (
        <IconOnly.type {...IconOnly.props} size={size === "2xl" ? 24 : 20} />
      ) : null}
      {TrailingIcon ? (
        <TrailingIcon.type
          {...TrailingIcon.props}
          className={classNames({
            "ml-2": size !== "2xl",
            "ml-3": size === "2xl",
          })}
        />
      ) : null}
    </button>
  );
};
