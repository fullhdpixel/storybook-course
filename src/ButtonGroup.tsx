import React, { FC } from "react";
import classNames from "classnames";

interface IButtonItem {
  value: string;
  content: React.ReactNode;
}

export interface ButtonGroupProps {
  active: string;
  setActive: (active: string) => void;
  options: IButtonItem[]; // Start with string[]
}

export const ButtonGroup: FC<ButtonGroupProps> = ({
  active,
  setActive,
  options,
}) => {
  return (
    <>
      {options.map((option, index) => {
        return (
          <button
            key={option.value}
            className={classNames("btn-group", {
              "rounded-l-lg border-r-0": index === 0,
              "border-r-0": index !== 0 && index !== options.length - 1,
              "rounded-r-lg": index === options.length - 1,
              "bg-gray-50 dark:bg-gray-700": active === option.value,
            })}
            onClick={() => setActive(option.value)}
          >
            {option.content}
          </button>
        );
      })}
    </>
  );
};
