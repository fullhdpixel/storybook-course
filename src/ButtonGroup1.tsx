import React from "react";
import classNames from "classnames";

interface IButtonItem {
  value: string;
  content: React.ReactNode;
}

export interface ButtonGroup1Props {
  active: string;
  setActive: (active: string) => void;
  options: IButtonItem[];
}

export const ButtonGroup1 = ({
  active,
  setActive,
  options,
}: ButtonGroup1Props) => {
  return (
    <>
      {options.map((option, index) => {
        return (
          <button
            key={option.value as string}
            className={classNames("btn-group", {
              "rounded-l-lg border-r-0": index === 0,
              "border-r-0": index !== 0 && index !== options.length - 1,
              "rounded-r-lg": index === options.length - 1,
              "bg-gray-50 dark:bg-gray-700": active === option.value,
            })}
            onClick={() => setActive(option.value)}
          >
            {option.content}
          </button>
        );
      })}
    </>
  );
};
