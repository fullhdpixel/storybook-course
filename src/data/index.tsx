export * from "./countries";
export * from "./dates";
export * from "./prices";
export * from "./rentals";
export * from "./navItems";
