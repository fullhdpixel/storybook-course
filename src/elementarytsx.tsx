// Basic types
// boolean, number, string, array, tuple, any, void

const isDone: boolean = false;
const age: number = 26;
const city: string = "Amsterdam";
const list: number[] = [1, 2, 3];
const list2: Array<number> = [4, 5, 6];
const coordinates: TCoord = [52.3825, 4.8064];

const getCoordinates = (): TCoord => coordinates;
const setCoordinates = (coordinates: TCoord) => console.log(coordinates);

// Custom Types
type TCoord = [number, number];

type BadgeSize = "sm" | "md" | "lg";
type BadgeVariant = "primary" | "secondary";

// Interfaces
interface BadgeProps {
  size: BadgeSize;
  variant: BadgeVariant;
}

// Extends
interface Props extends BadgeProps {
  darkMode: boolean;
}

// Generic Interfaces
interface ButtonGroupProps<T> {
  active: T;
  setActive: (active: T) => void;
}

// Record
const BadgeClasses: Record<BadgeSize, string> = {
  sm: "badge-sm",
  md: "badge-md",
  lg: "badge-lg",
};
