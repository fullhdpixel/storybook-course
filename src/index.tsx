export * from "./Badge";
export * from "./Button";
export * from "./ButtonGroup";
export * from "./Select";
export * from "./Paginate";
export * from "./Typography";
export * from "./Navbar";
export * from "./Input";
export * from "./RentalCard";
export * from "./MobileNavbar";
