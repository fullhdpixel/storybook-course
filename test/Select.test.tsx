import * as React from "react";
import {
  fireEvent,
  render,
  cleanup,
  RenderResult,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { countries, prices, dates, Select, SelectProps } from "../src";

afterEach(cleanup);

interface ISetupSelect extends RenderResult {
  select: HTMLSelectElement;
}

const setup = ({
  options,
  selectedOption,
  setSelectedOption,
  placeholder = "Select an option",
  width = "w-50",
}: SelectProps): ISetupSelect => {
  const utils = render(
    <Select
      options={options}
      selectedOption={selectedOption}
      setSelectedOption={setSelectedOption}
      placeholder={placeholder}
      width={width}
    />,
  );

  const select = utils.getByTestId("selectContainer") as HTMLSelectElement;

  return {
    select,
    ...utils,
  };
};

// TODO try to create snapshot of opened dropdown
describe("Select", () => {
  it("Renders CountrySelect correctly", () => {
    const { asFragment } = setup({
      options: countries,
      selectedOption: countries[0],
      setSelectedOption: () => {},
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders PriceSelect correctly", () => {
    const { asFragment } = setup({
      options: prices,
      selectedOption: undefined,
      placeholder: "Select a price",
      setSelectedOption: () => {},
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders DateSelect correctly", () => {
    const { asFragment } = setup({
      options: dates,
      selectedOption: dates[0],
      setSelectedOption: () => {},
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("Selecting an option works", () => {
    const setSelectedOption = jest.fn();

    const { getByText, getByTestId } = setup({
      options: countries,
      selectedOption: countries[0],
      setSelectedOption,
    });
    const { value } = countries[4];

    // Click on button
    fireEvent.click(getByTestId("selectButton"));
    // Panel opens, select US
    fireEvent.click(getByText("Washington, US"));

    expect(setSelectedOption).toHaveBeenCalledWith(value);
  });
});
