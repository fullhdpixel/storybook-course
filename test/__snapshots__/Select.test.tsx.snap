// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Renders CountrySelect correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block w-50"
    data-testid="selectContainer"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="shadow-sm flex items-center text-md border border-gray-300 dark:border-gray-500 h-11 px-3.5 rounded-lg bg-white dark:bg-gray-800 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 whitespace-nowrap text-gray-900 dark:text-white w-50"
      data-testid="selectButton"
      id="headlessui-listbox-button-1"
      type="button"
    >
      <img
        alt="nl"
        class="w-5 h-5 mr-2"
        src="https://res.cloudinary.com/tailwindcss/image/upload/v1635279277/nl_tpy2ab.svg"
      />
       Amsterdam, NL
      <svg
        class="text-gray-500 dark:text-gray-300 transform duration-100 ease-out ml-auto"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="6 9 12 15 18 9"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Select Renders DateSelect correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block w-50"
    data-testid="selectContainer"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="shadow-sm flex items-center text-md border border-gray-300 dark:border-gray-500 h-11 px-3.5 rounded-lg bg-white dark:bg-gray-800 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 whitespace-nowrap text-gray-900 dark:text-white w-50"
      data-testid="selectButton"
      id="headlessui-listbox-button-5"
      type="button"
    >
      Okt 25 - Okt 31
      <svg
        class="text-gray-500 dark:text-gray-300 transform duration-100 ease-out ml-auto"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="6 9 12 15 18 9"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Select Renders PriceSelect correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block w-50"
    data-testid="selectContainer"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="shadow-sm flex items-center text-md border border-gray-300 dark:border-gray-500 h-11 px-3.5 rounded-lg bg-white dark:bg-gray-800 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 whitespace-nowrap text-gray-900 dark:text-white w-50"
      data-testid="selectButton"
      id="headlessui-listbox-button-3"
      type="button"
    >
      Any price
      <svg
        class="text-gray-500 dark:text-gray-300 transform duration-100 ease-out ml-auto"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="6 9 12 15 18 9"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
